akka-stream/src/main/scala/akka/stream/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,212 @@
+
+      override def preStart(): Unit = {
+        self = getStageActor(initialReceive)
+        if (initialPartnerRef.isDefined) // this will set the `partnerRef`
+          observeAndValidateSender(initialPartnerRef.get, "Illegal initialPartnerRef! This would be a bug in the SinkRef usage or impl.")
+
+        log.debug("Created SinkRef, pointing to remote Sink receiver: {}, local worker: {}", initialPartnerRef, self.ref)
+
+        if (canMaterializeSourceRef)
+          promise.success(new SourceRef(OptionVal(self.ref), canMaterializeSinkRef = false))
+
+        if (partnerRef.isDefined) {
+          getPartnerRef ! StreamRefs.OnSubscribeHandshake(self.ref)
+          tryPull()
+        }
+
+        scheduleOnce(SubscriptionTimeoutTimerKey, subscriptionTimeout.timeout)
+      }
+
+      lazy val initialReceive: ((ActorRef, Any)) ⇒ Unit = {
+        case (_, Terminated(ref)) ⇒
+          if (ref == getPartnerRef)
+            failStage(RemoteStreamRefActorTerminatedException(s"Remote target receiver of data $partnerRef terminated. " +
+              s"Local stream terminating, message loss (on remote side) may have happened."))
+
+        case (sender, CumulativeDemand(d)) ⇒
+          observeAndValidateSender(sender, "Illegal sender for CumulativeDemand")
+
+          if (remoteCumulativeDemandReceived < d) {
+            remoteCumulativeDemandReceived = d
+            log.debug("Received cumulative demand [{}], consumable demand: [{}]", CumulativeDemand(d), remoteCumulativeDemandReceived - remoteCumulativeDemandConsumed)
+          }
+
+          tryPull()
+      }
+
+      override def onPush(): Unit = {
+        val elem = grabSequenced(in)
+        getPartnerRef ! elem
+        log.debug("Sending sequenced: {} to {}", elem, getPartnerRef)
+        tryPull()
+      }
+
+      private def tryPull() =
+        if (remoteCumulativeDemandConsumed < remoteCumulativeDemandReceived && !hasBeenPulled(in)) {
+          pull(in)
+        }
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case SubscriptionTimeoutTimerKey ⇒
+          val ex = StreamRefs.StreamRefSubscriptionTimeoutException(
+            // we know the future has been competed by now, since it is in preStart
+            s"[$stageActorName] Remote side did not subscribe (materialize) handed out Sink reference [${promise.future.value}]," +
+              s"within subscription timeout: ${PrettyDuration.format(subscriptionTimeout.timeout)}!")
+
+          throw ex // this will also log the exception, unlike failStage; this should fail rarely, but would be good to have it "loud"
+      }
+
+      private def grabSequenced[T](in: Inlet[T]): SequencedOnNext[T] = {
+        val onNext = SequencedOnNext(remoteCumulativeDemandConsumed, grab(in))
+        remoteCumulativeDemandConsumed += 1
+        onNext
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit =
+        if (partnerRef.isDefined) {
```

Comment:
why not `partnerRef match` instead of these isDefined / get ?

Reply:
because `OptionVal`